File: Generated_Codes_Check/Results/python/92.py
"""

Explanation
In the first query (0 2 5), the minimum integer in the range [2, 5] is 2. In the second query (1 1 5 6), the elements of the array become A = [1, 5, 6, 6, 4]. In the third query (0 2 2), the minimum integer in the range [2, 2] is 6. In the fourth query (1 2 5 3), the elements of the array become A = [1, 5, 2, 2, 2]. In the fifth query (0 1 3), the minimum integer in the range [1, 3] is 2.

"""

def createSegTree(arr,tree,start,end,index):
#     print(arr,tree,start,end,index)
    if start == end:
        tree[index] = arr[start]
        return
    mid = (start+end)//2
    createSegTree(arr,tree,start,mid,2*index)
    createSegTree(arr,tree,mid+1,end,2*index+1)
    tree[index] = min(tree[2*index],tree[2*index+1])
#     print( tree[index])
    
def query(tree,start,end,index,l,r):
#     print(tree,start,end,index,l,r)
    if r<start or l>end or start>end:
        return 999999999999999
    if start>=l and end<=r:
        return tree[index]
    mid = (start+end)//2
    return min(query(tree,start,mid,2*index,l,r),query(tree,mid+1,end,2*index+1,l,r))
    
def updateTree(arr,tree,start,end,index,l,r,x):
    if start>end or l>end or r<start:
        return
    if start == end:
        tree[index] = arr[start]
        return
    mid = (start+end)//2
    updateTree(arr,tree,start,mid,2*index,l,r,x)
    updateTree(arr,tree,mid+1,end,2*index+1,l,r,x)
    tree[index] = min(tree[2*index],tree[2*index+1])
    
def update(arr,tree,start,end,index,pos,value):
    if start == end:
        arr[pos] = value
        tree[index] = value
        return
    mid = (start+end)//2
    if pos>mid:
        update(arr,tree,mid+1,end,2*index+1,pos,value)
    else:
        update(arr,tree,start,mid,2*index,pos,value)
    tree[index] = min(tree[2*index],tree[2*index+1])
    
 

def main():
    n,q = map(int,input().strip().split())
    arr = list(map(int,input().strip().split()))
    tree = [0]*(4*n)
    createSegTree(arr,tree,0,n-1,1)
    for i in range(q):
        querytype = list(map(int,input().strip().split()))
#         print(querytype)
        if querytype[0] == 0:
            ans = query(tree,0,n-1,1,querytype[1],querytype[2])
            print(ans)
        elif querytype[0] == 1:
            updateTree(arr,tree,0,n-1,1,querytype[1],querytype[2],querytype[3])
        else:
            update(arr,tree,0,n-1,1,querytype[1],querytype[2])
main()

File: Generated_Codes_Check/Results/java/92.java
/*


*/
import java.util.*;
import java.io.*;

class Main {
  public static void main(String[] args) throws IOException {
    Scanner in = new Scanner(System.in);
    int N = in.nextInt();
    int Q = in.nextInt();
    int[] A = new int[N];
    for(int i = 0; i < N; i++)
        A[i] = in.nextInt();
    SegmentTree st = new SegmentTree(A,N);
    StringBuilder sb = new StringBuilder();
    while(Q > 0) {
        int query = in.nextInt();
        int L = in.nextInt();
        int R = in.nextInt();
        if(query == 0) {
            sb.append(st.min(L-1,R-1,1,0,N-1)).append("\n");
        } else {
            int X = in.nextInt();
            st.update(L-1,R-1,X,1,0,N-1);
        }
        Q--;
    }
    System.out.print(sb);
  }
}

class SegmentTree {
    int[] tree;
    int[] A;
    SegmentTree(int[] arr, int n) {
        int h = (int)Math.ceil(Math.log(n)/Math.log(2));
        int max = 2*(int)Math.pow(2,h)-1;
        tree = new int[max];
        A = new int[n];
        for(int i = 0; i < n; i++)
            A[i] = arr[i];
        construct(0,n-1,1);
    }
    int construct(int ss, int se, int index) {
        if(ss == se) {
            tree[index] = A[ss];
            return A[ss];
        }
        int mid = (ss+se)/2;
        int left = construct(ss,mid,2*index);
        int right = construct(mid+1,se,2*index+1);
        tree[index] = Math.min(left,right);
        return tree[index];
    }
    int min(int qs, int qe, int index, int ss, int se) {
        if(qs <= ss && qe >= se)
            return tree[index];
        if(qe < ss || qs > se)
            return Integer.MAX_VALUE;
        int mid = (ss+se)/2;
        int left = min(qs,qe,2*index,ss,mid);
        int right = min(qs,qe,2*index+1,mid+1,se);
        return Math.min(left,right);
    }
    void update(int qs, int qe, int newval, int index, int ss, int se) {
        if(qs <= ss && qe >= se) {
            tree[index] &= newval;
            return;
        }
        if(qe < ss || qs > se)
            return;
        int mid = (ss+se)/2;
        update(qs,qe,newval,2*index,ss,mid);
        update(qs,qe,newval,2*index+1,mid+1,se);
        tree[index] = Math.min(tree[2*index],tree[2*index+1]);
    }
}